src/types/sealedClass/Rationale.md

```diff
@@ -0,0 +1,9 @@
+Correct answer: **a) Yes**
+
+* Exhaustive check on sealed classes does not happen when it is checked in a statement. Convert the check into expression as:- 
+```
+val out = when (h) {
+    is Dog -> println("It is a Dog")
+}
+```
```

Comment:
note: statements can be made exhaustive in another way too: call a function or property on them
```
when (h) {
    is Animal.Dog -> println("It is a Dog")
}.javaClass
```
although this looks tricky and unrelated, it's better to use this really nice addition to Kotlin
```
val <T> T.exhaustive: T get() = this
```
from the blog post below

Reply:
Agreed `val <T> T.exhaustive: T get() = this` is much nicer as `javaclass` doesn't tell much about the reason for use. However I really don't like adding extension function to every type. We can also convert statements to expressions by passing them as lambda. e.g. ->
```
fun <T> exhaustive(func: () -> T) {
    func()
}

fun test_sealed_class_correct() {
    val h = Human() as Animal
    exhaustive {
        when (h) {
            is Animal.Dog -> null
        //Gives error!
        }
    }
}

```

Reply:
Nice idea as well, thanks for sharing; interesting that even though the return value is apparently (apart from being last expression in lambda) not used, it's still exhaustive. I agree about polluting the namespace.

Note there's also this: https://youtrack.jetbrains.com/issue/KT-12380

Just a bonkers idea for fun, cause nesting is ugly:
```
@Suppress("ClassName", "NOTHING_TO_INLINE")
object an {
    inline infix fun <T> exhaustive(whenValue: T) = whenValue
}

fun test_sealed_class_correct() {
    val h = Human() as Animal
    an exhaustive when (h) {
        is Animal.Dog -> null
        //Gives error!
    }
}
```